lib/allocate.rs

```diff
@@ -1,39 +1,78 @@
+extern crate num;
 use std::collections::BTreeSet;
+use std::ops;
+use std::cmp;
 
-pub fn allocate_allocate_u64<B: Ord+Clone>(allocated: &std_Set<u64>, toallocate: &std_Vec<B>, max_val: &u64) -> std_Vec<(B,u64)> {
-    let first = *allocated.x.range(..).next_back().unwrap_or(&0);
-    let mut last = first;
+pub fn allocate_allocate<B: Ord+Clone, N:num::Num+ops::Add+cmp::Ord+Copy>(
+    allocated: &std_Set<N>,
+    toallocate: &std_Vec<B>,
+    min_val: &N,
+    max_val: &N) -> std_Vec<(B,N)>
+{
+    assert!(*max_val >= *min_val);
+    let range = *max_val - *min_val;
+
+    // Next index to consider
+    let mut next = *allocated.x.range(..).next_back().unwrap_or(&max_val);
+    // allocated may contain values outside of the [min_val..max_val] range.
+    if next < *min_val || next > *max_val {
+        next = *max_val;
+    };
+    let mut offset = N::zero();
     let mut res = std_Vec::new();
     for b in toallocate.x.iter() {
         loop {
-            last = if last + 1 <= *max_val {last + 1} else {1};
-            if last == first {
-                return res;
-            };
-            if allocated.x.contains(&last) { continue; };
-            res.x.push(((*b).clone(), last));
-            break;
+            next = if next == *max_val { *min_val } else { next + N::one() };
```

Comment:
won't this get into an infinite loop if all values are allocated?

Reply:
It uses `offset` to terminate the loop, so should be fine.